src/charm.py

```diff
@@ -138,17 +146,21 @@ def _reconfigure(self, event: ops.charm.RelationEvent) -> None:
             event: The triggering relation event.
         """
         # check if reconfiguration is necessary
-        if self._need_replica_set_reconfiguration:
-            try:
-                self._mongo.reconfigure_replica_set()
-                # update the set of replica set hosts
-                self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
-                logger.debug("Replica set successfully reconfigured")
-                self.unit.status = ActiveStatus()
-            except (ConnectionFailure, ConfigurationError, OperationFailure) as e:
-                logger.error("deferring reconfigure of replica set: %s", str(e))
-                self.unit.status = WaitingStatus("waiting to reconfigure replica set")
-                event.defer()
+        if not self._need_replica_set_reconfiguration:
+            self.unit.status = ActiveStatus()
+            return
+
+        try:
+            self._mongo.reconfigure_replica_set()
+            # update the set of replica set hosts
+            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
+            logger.debug("Replica set successfully reconfigured")
+            self.unit.status = ActiveStatus()
+        except (ConnectionFailure, ConfigurationError, OperationFailure) as e:
+            logger.error("deferring reconfigure of replica set: %s", str(e))
+            self.unit.status = WaitingStatus("waiting to reconfigure replica set")
+            event.defer()
+            return
```

Comment:
minor minor nit: may be these `event.defer()` and `return` are no longer necessary?

Reply:
Great point! `return` is definitely no longer needed, but `event.defer()` is since we need to re-trigger the event, updating soon :D